procmon_parser/symbol_resolver/symbol_resolver.py

```diff
@@ -0,0 +1,711 @@
+            # The following 2 lines just do that.
+            file_name = ctypes.create_unicode_buffer(line.FileName)  # noqa
+            line.FileName = ctypes.cast(file_name, ctypes.c_wchar_p)
+
+            logger.debug("File Name: '{line.FileName}'; Line Number: {line.LineNumber}; "
+                         "Line Displacement (col): {line_displacement.value}".format(
+                         line=line, line_displacement=line_displacement))  # noqa
+
+            # It's possible that the returned line.Filename is already a fully qualified path, in which case there's no
+            #    need to call SymGetSourceFileW, as the latter would be only used to retrieve the fully qualified path.
+            # We just check that we already have fully qualified path. If it is, then we bail out, otherwise we call
+            #    SymGetSourceFileW.
+            if pathlib.Path(line.FileName).is_absolute():  # noqa
+                # we have a fully qualified source file path.
+                logger.debug("source file path [from line.Filename]: {line.FileName}".format(line=line))
+                fully_qualified_source_path = line.FileName
+            else:
+                # we don't have a fully qualified source file path.
+                source_file_path_size = DWORD(260)
+                source_file_path = ctypes.create_unicode_buffer(source_file_path_size.value)
+                ret_val = self._dbghelp.SymGetSourceFileW(
+                    pid,  # hProcess
+                    module.base_address,  # Base
+                    None,  # Params (never used)
+                    line.FileName,  # FileSpec (name of source file) [PCWSTR]
+                    source_file_path,  # [out] FilePath: fully qualified path of source file
+                    source_file_path_size  # FilePath size (num chars)
+                )
+                if ret_val == 0:
+                    last_err = ctypes.get_last_error()
+                    logger.debug("SymGetSourceFileW: ({last_err:#08x}) {formatted_last_err}".format(
+                        last_err=last_err, formatted_last_err=ctypes.FormatError(last_err)))
+                    logger.debug("SymGetSourceFileW failed: using '{line.FileName}' as fallback.".format(
+                        line=line))
+                    # use line.FileName as fallback
+                    fully_qualified_source_path = line.FileName
+                else:
+                    logger.debug("source file path [from SymGetSourceFileW]: {source_file_path.value}".format(
+                        source_file_path=source_file_path))
+                    fully_qualified_source_path = source_file_path.value
+
+            yield StackTraceFrameInformation(frame_type, frame_number, address, module, symbol_info, displacement.value,
+                                             line, line_displacement.value, fully_qualified_source_path)
+
+        # dbghelp symbol cleanup
+        self._dbghelp.SymCleanup(pid)
```

Comment:
using [contextlib ](https://docs.python.org/3/library/contextlib.html) is more safe so `SymCleanup` will be called in case an error is raised inside the function

Reply:
I tried to think about that, but don't know how to implement it (I mean I know how to implement a context manager in Python, but don't know how to do it in this case).

[`SymInitialize`](https://learn.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-syminitialize) call is tied to a unique id (first parameter) which then must be passed to nearly all symbol functions. As of now the call is in the the `resolve_stack_trace` as I decided to tie it to the PID of the process for which the stack trace is to be retrieved. Basically, each event - for which the stack trace is to resolved - is getting a call to `SymInitialize`. That's what debuggers do. 
As of now, it's not really possible to have a context manager to call `SymCleanup` (because it's a function exit).

1. Now, I don't know if moving the call (`SymInitialize`) to the `__init__` with a bogus - but unique - value would work (e.g. 0xDEADBEEF or something like that). In which case the context manager would not be a problem to implement. Not very clear to me what this unique value is used for internally in dbghelp.

2. Another possibility would be to move everything to an internal function and try catch on the public facing function, I guess something like that?:

```python
def resolve_stack_trace(self, event):
    try:
        def _resolve_stack_trace(event)  # call internal function which does the real job.
    except Exception:
        raise
    finally:
        self._dbghelp.SymCleanup(pid)  # need to keep track of the pid in __init__.
```

What's your view on that?

Reply:
I also didn't understand what is the intended pid value to use in `SymInitialize` and if it's better to initialize once or for every event... in either case I think we should use the second way you proposed, you can do that with `contextlib` with something like

```python
@contextlib.contextmanager
    def symsrv_open(self, event):
        pid = event.process.pid
        result = self._dbghelp.SymInitialize(pid, None, False)
        if not result:
            raise RuntimeError("Error SymInitialize {}".format(result))
        try:
            yield pid
        finally:
            self._dbghelp.SymCleanup(pid)
```

and then

```python
with self.symsrv_open(event) as sym_pid:
    self._resolve_stack_trace(event)
```

Reply:
Thanks a lot for your feedback! iirc Procmon does call `SymInitialize` for **each** event, with the pid of the process as the first argument to the call. I'll try to implement something following your guidelines.